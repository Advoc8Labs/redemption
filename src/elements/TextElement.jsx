import React from 'react';
import Delta from 'quill-delta';
import generateNodes from '../components/TextNodes';

function TextElement({ element, values, handleRedirect }) {
    const nodes = generateNodes({
        element,
        values,
        handleRedirect,
        delta: new Delta(element.text_formatted),
        elementID: element.id
    });

    const styles = element.applyStyles;
    return (
        <div
            css={{
                display: 'flex',
                flexDirection: 'column',
                ...styles.getLayout(),
                ...styles.getTarget('container')
            }}
        >
            <div css={styles.getTarget('text')}>{nodes}</div>
        </div>
    );
}

export default TextElement;
