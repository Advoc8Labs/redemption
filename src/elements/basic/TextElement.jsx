import React, { useMemo } from 'react';
import TextNodes from '../components/TextNodes';

// TODO(peter): deprecate once customers have upgraded and backend migrated
function legacyAlignment(alignment) {
  switch (alignment) {
    case 'flex-start':
      return 'left';
    case 'flex-end':
      return 'right';
    default:
      return alignment;
  }
}

function applyTextStyles(element, applyStyles) {
  applyStyles.addTargets('text');
  applyStyles.apply('text', 'layout', (a) => ({
    textAlign: legacyAlignment(a)
  }));
  applyStyles.applyMargin('text');
  if (element.styles.line_height) {
    applyStyles.apply('text', 'line_height', (a) => ({
      lineHeight: `${a}px`
    }));
  }
  return applyStyles;
}

function TextElement({
  element,
  applyStyles,
  values = null,
  handleRedirect = () => {},
  conditions = []
}) {
  const styles = useMemo(() => applyTextStyles(element, applyStyles), [
    applyStyles
  ]);
  return (
    <div css={styles.getTarget('text')}>
      <TextNodes
        element={element}
        values={values}
        applyStyles={applyStyles}
        handleRedirect={handleRedirect}
        conditions={conditions}
      />
    </div>
  );
}

export default TextElement;
