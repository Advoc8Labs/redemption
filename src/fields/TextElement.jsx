import React from 'react';
import Delta from 'quill-delta';
import generateNodes from './textNodes';
import { alignmentMap } from '../utils/formHelperFunctions';

/**
 * Disambiguation: this is NOT a text "field" that receives input.
 * It just models a block of text in a form.
 */
function TextElement({ field, fieldValues, conditions, submit }) {
    const elementID = field.id;
    let delta = new Delta(field.text_formatted);

    conditions.forEach((cond) => {
        if (cond.element_type === 'text' && cond.element_id === elementID) {
            const start = cond.metadata.start || 0;
            const end = cond.metadata.end || field.text.length;
            delta = delta.compose(
                new Delta().retain(start).retain(end - start, { start, end })
            );
        }
    });

    const repeat = field.repeat || 0;
    const nodes = generateNodes({
        delta,
        fieldValues,
        field,
        submit,
        repeat,
        elementID
    });
    const style = {
        gridColumnStart: field.column_index + 1,
        gridRowStart: field.row_index + 1,
        gridColumnEnd: field.column_index_end + 2,
        gridRowEnd: field.row_index_end + 2,
        paddingBottom: `${field.padding_bottom}px`,
        paddingTop: `${field.padding_top}px`,
        paddingLeft: `${field.padding_left}px`,
        paddingRight: `${field.padding_right}px`,
        display: 'flex',
        flexDirection: 'column',
        alignItems: alignmentMap[field.layout],
        textAlign: field.layout,
        justifyContent: field.vertical_layout
    };
    if (field.border_color) style.border = `1px solid #${field.border_color}`;

    return (
        <div key={field.id} css={style}>
            <div>{nodes}</div>
        </div>
    );
}

export default TextElement;
