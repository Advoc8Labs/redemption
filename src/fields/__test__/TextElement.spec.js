import React from 'react';
import { create, act } from 'react-test-renderer';
import TextElement from '../TextElement';

describe('Text', () => {
    it('renders an empty text block', async () => {
        // Arrange
        const props = {
            field: { text_formatted: [{ insert: '\n' }] },
            fieldValues: {},
            conditions: [],
            displaySteps: [],
            submit: () => {},
            setElementKey: () => {},
            setRepeat: () => {}
        };

        // Act
        let text;
        act(() => {
            text = create(<TextElement {...props} />);
        });
        const tree = text.toJSON();

        // Assert
        expect(tree).toMatchSnapshot();
    });

    it('renders a plaintext text block', async () => {
        // Arrange
        const props = {
            field: { text_formatted: [{ insert: 'Hello World!' }] },
            fieldValues: {},
            conditions: [],
            displaySteps: [],
            submit: () => {},
            setElementKey: () => {},
            setRepeat: () => {}
        };

        // Act
        let text;
        act(() => {
            text = create(<TextElement {...props} />);
        });
        const tree = text.toJSON();

        // Assert
        expect(tree).toMatchSnapshot();
    });

    it('renders a text block with custom styles', async () => {
        // Arrange
        const props = {
            field: {
                text_formatted: [
                    {
                        insert: 'Hello World!',
                        attributes: {
                            weight: 700,
                            color: '#000000FF',
                            size: '48px',
                            font: 'sans-serif',
                            italic: true,
                            strike: false,
                            underline: false
                        }
                    }
                ]
            },
            fieldValues: {},
            conditions: [],
            displaySteps: [],
            submit: () => {},
            setElementKey: () => {},
            setRepeat: () => {}
        };

        // Act
        let text;
        act(() => {
            text = create(<TextElement {...props} />);
        });
        const tree = text.toJSON();

        // Assert
        expect(tree).toMatchSnapshot();
    });
});
