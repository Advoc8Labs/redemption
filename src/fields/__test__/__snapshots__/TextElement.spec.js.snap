// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text renders a plaintext text block 1`] = `
<div
  className="css-kevpsq-TextElement"
>
  <div
    style={
      Object {
        "marginBottom": "undefinedpx",
        "marginLeft": "undefinedpx",
        "marginRight": "undefinedpx",
        "marginTop": "undefinedpx",
      }
    }
  >
    <span
      onClick={[Function]}
      style={
        Object {
          "whiteSpace": "pre-wrap",
        }
      }
    >
      Hello World!
    </span>
  </div>
</div>
`;

exports[`Text renders a text block with custom styles 1`] = `
<div
  className="css-kevpsq-TextElement"
>
  <div
    style={
      Object {
        "marginBottom": "undefinedpx",
        "marginLeft": "undefinedpx",
        "marginRight": "undefinedpx",
        "marginTop": "undefinedpx",
      }
    }
  >
    <span
      onClick={[Function]}
      style={
        Object {
          "color": "##000000FF",
          "fontSize": "48pxpx",
          "fontStyle": "italic",
          "fontWeight": 700,
          "whiteSpace": "pre-wrap",
        }
      }
    >
      Hello World!
    </span>
  </div>
</div>
`;

exports[`Text renders an empty text block 1`] = `
<div
  className="css-kevpsq-TextElement"
>
  <div
    style={
      Object {
        "marginBottom": "undefinedpx",
        "marginLeft": "undefinedpx",
        "marginRight": "undefinedpx",
        "marginTop": "undefinedpx",
      }
    }
  >
    <span
      onClick={[Function]}
      style={
        Object {
          "whiteSpace": "pre-wrap",
        }
      }
    >
      

    </span>
  </div>
</div>
`;
